dataclients/kubernetes/definitions/routegroupvalidator.go

```diff
@@ -0,0 +1,335 @@
+		}
+		names[br.BackendName] = struct{}{}
+
+		if err := br.validate(backends); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (br *BackendReference) validate(backends map[string]bool) error {
+	if br == nil || br.BackendName == "" {
+		return errUnnamedBackendReference
+	}
+
+	if !backends[br.BackendName] {
+		return invalidBackendReference(br.BackendName)
+	}
+
+	if br.Weight < 0 {
+		return invalidBackendWeight(br.BackendName, br.Weight)
+	}
+
+	return nil
+}
+
+func (r *RouteSpec) UniqueMethods() []string {
+	return uniqueStrings(r.Methods)
+}
+
+// TODO: we need to pass namespace/name in all errors
+func (rg *RouteGroupSpec) validate() error {
+	// has at least one backend:
+	if len(rg.Backends) == 0 {
+		return errRouteGroupWithoutBackend
+	}
+
+	// backends valid and have unique names:
+	backends := make(map[string]bool)
+	for _, b := range rg.Backends {
+		if backends[b.Name] {
+			return backendsWithDuplicateName(b.Name)
+		}
+
+		backends[b.Name] = true
+		if err := b.validate(); err != nil {
+			return invalidBackend(b.Name, err)
+		}
+	}
+
+	hasDefault := len(rg.DefaultBackends) > 0
+	if err := rg.DefaultBackends.validate(backends); err != nil {
+		return err
+	}
+
+	if !hasDefault && len(rg.Routes) == 0 {
+		return errMissingBackendReference
+	}
+
+	for i, r := range rg.Routes {
+		if err := r.validate(hasDefault, backends); err != nil {
+			return invalidRoute(i, err)
+		}
+	}
+
+	return nil
+}
+
+// TODO: we need to pass namespace/name in all errors
+func (r *RouteSpec) validate(hasDefault bool, backends map[string]bool) error {
+	if r == nil {
+		return errInvalidRouteSpec
+	}
+
+	if !hasDefault && len(r.Backends) == 0 {
+		return errMissingBackendReference
+	}
+
+	if err := r.Backends.validate(backends); err != nil {
+		return err
+	}
+
+	if r.Path != "" && r.PathSubtree != "" {
+		return errBothPathAndPathSubtree
+	}
+
+	if hasEmpty(r.Predicates) {
+		return errInvalidPredicate
+	}
+
+	if hasEmpty(r.Filters) {
+		return errInvalidFilter
+	}
+
+	if hasEmpty(r.Methods) {
+		return errInvalidMethod
+	}
+
+	return nil
+}
+
+// TODO: we need to pass namespace/name in all errors
+func (r *RouteGroupItem) validate() error {
+	// has metadata and name:
+	if r == nil || validate(r.Metadata) != nil {
+		return errRouteGroupWithoutName
+	}
+
+	// has spec:
+	if r.Spec == nil {
+		return routeGroupError(r.Metadata, errRouteGroupWithoutSpec)
+	}
+
+	if err := r.Spec.validate(); err != nil {
+		return routeGroupError(r.Metadata, err)
+	}
+
+	return nil
+}
+
+// ParseRouteGroupsJSON parses a json list of RouteGroups into RouteGroupList
+func ParseRouteGroupsJSON(d []byte) (RouteGroupList, error) {
+	var rl RouteGroupList
+	err := json.Unmarshal(d, &rl)
+	return rl, err
+}
+
+// ParseRouteGroupsYAML parses a YAML list of RouteGroups into RouteGroupList
+func ParseRouteGroupsYAML(d []byte) (RouteGroupList, error) {
+	var rl RouteGroupList
+	err := yaml.Unmarshal(d, &rl)
+	return rl, err
+}
+
+type RouteGroupValidator struct {
```

Comment:
Thought about making `Validator` interface, something like

``` go
type Validator interface {
          Validate(item interface{}) error
}
```

but I think it's to early for this. It was to make all theses `validate` functions to be used in other places and `interface{}` wasn't the best thing to use in this place in my opinion.

Reply:
You can also use an interface function that doesn't pass a value as func argument.
Then you can use proper OO to figure out how to implement the `Validate() error` interface for a given type.